### DISCUSSION TITLE
IsRequired support for Relationship field

### DISCUSSION THREAD

**Discussion by SeanDoyleGit** [keystonejs/keystone#7310] [Feature requests]
# Bug report
The `IsRequired` property on the `Relationship` field  is not enforced.

## Describe the bug

An item should not be created / updated if there is an empty `Relationship` field  with an `IsRequired` property set to true.

## To Reproduce

Open the blog demo example and make the following changes. 

1. Add the `isRequired` property to the `categories` field  on the Post list
```
categories: {
  type: Relationship,
  ref: 'PostCategory',
  many: true,
  isRequired: true,
},
```

2. Run the blog demo and go to the Post list page http://localhost:3000/admin/posts?fields=_label_

3. Click the create button and and fill out the Create Post form, make sure to leave the categories field blank. 

4. Click create and you will see the post is created without any categories. 

## Expected behaviour

An error should occur preventing the creation of the post as the category field is empty.

**Comment by NoTuxNoBux** [comment]
For what it's worth: this issue seems to have been posted for Keystone 5, but it still applies to 6 or Next. In fact, `relationship` doesn't appear to have _any_ `isRequired` property anymore - TypeScript highlights this as an error.

(I'm mentioning this instead of upvoting because the ticket is marked as needing review, likely because it wasn't certain if it still applied to Keystone Next.)

**Comment by timur-safarov** [comment]
Why owners still don't solve this bug - that's an awful mistake.

**Comment by molomby** [comment]
Cribbing from my recent [SO answer](https://stackoverflow.com/a/71419431/762028) on this topic:

> Relationship fields currently don't support the `validation.isRequired` or `db.isNullable` options. This is true even when the list being configured holds the foreign key (ie. a many-to-one relationship or one-to-one with `db.foreignKey: true`).
> 
> There are plans to support these options but the work isn't trivial. For example, these constraints can effect the order in which nested creates need to be performed. Keystone will also need to validate the config makes sense and doesn't for example, have `isNullable: false` on both sides of a one-to-one relationship (which would make inserting records impossible).
> 
> If you want to emulate similar functionality right now it's possible using [hooks](https://keystonejs.com/docs/guides/hooks). I think you'd need...
> 
> * A [`validate-input`](https://keystonejs.com/docs/apis/hooks#before-operation) hook on the list with the foreign key, to ensure a item was linked on create (and not removed on update)
> * [`validate-input`](https://keystonejs.com/docs/apis/hooks#before-operation) and [`validate-delete`](https://keystonejs.com/docs/apis/hooks#validate-delete) hooks on the other list to ensure links weren't broken when updating or deleting items from the other side.
> 
> Since this solution's implemented in the app layer it doesn't give you as strong a guarantee as a proper database constraint, but it's a start.

**Reply by willemmulder** [reply]
@molomby Could you elaborate on how validation.isRequired yields extra constraints on (nested) creates?

Edit: was just thinking that that probably happens when the relationship is also a new record created in the admin UI, meaning you'd have to store records 'bottom-to-top'. Requiring an algorithm to figure out the order-of-insertion. And yes, the problem would be circular dependencies that make inserting records impossible... But the same algorithm that determines the order-of-insertion would hopefully/probably also be able to detect circular dependencies upfront, so these schema issues can be detected at build time.

**Reply by molomby** [reply]
Pretty much, yeah. It's certainly possible, just adds a bunch of complexity that Keystone can currently ignore.

It's been a while since I looked at this but I think it also complicates how we need to use underlying Prisma API..? Keystone tries to wrap Prisma fairly lightly and I suspect fixing this well would require us to reach in a little deeper than we'd like. @emmatown would know.

The other approach (that I favour) is to let people configure not-nullable on foreign keys then, if it causes the integrity constraints cause some nested create to fail, just error back to the user. It's a minor foot gun but, then again, if you're doing complex nested operations, maybe best to build a custom mutation anyway? The CRUD API isn't intended to solve for all use cases and, right now, this limitation blocks devs from building the DB schema they need.

**Reply by willemmulder** [reply]
Yes exactly; that would be great. I.e. for now just allow isNullable:false and error out when problems arise, and then we can possibly improve it from there? Also interested what @emmatown has to say about this!

**Comment by jschuur** [comment]
The main problem with a `validateInput` hooks based approach at checking for required relationships is that `resolvedData` will happily tell you how many entries were deleted, but it won't tell you the current count that exist to compare that against. This makes it harder to find out if all the entries were removed in an update action.

I came up with the following solution, that compares the number of deleted entries via a Prisma query with the current count for hasMany relationships and also works for hasOne:

```js
async function hasManyCheck({
  operation,
  field,
  listKey,
  context,
  resolvedData,
  item,
  addValidationError,
}) {
  if (operation === 'create' && !resolvedData[field])
    addValidationError(`Missing required relationship: ${field}`);
  else {
    const currentCount =
      (
        await context.prisma[listKey].findFirst({
          where: {
            id: { equals: item?.id },
          },
          select: {
            _count: {
              select: {
                [field]: true,
              },
            },
          },
        })
      )?._count?.[field] || 0;
    const deletedCount = resolvedData?.[field]?.disconnect?.length || 0;

    if (currentCount && currentCount === deletedCount)
      addValidationError(`Missing required relationship: ${field}`);
  }
}

const hasOneExistingValue = async ({ field, item, listKey, context }: any) =>
  (await context.prisma[listKey].count({
    where: {
      id: { equals: item?.id },
      [field]: { isNot: null },
    },
  })) > 0;

export async function validateRelationships({ hasOne = [], hasMany = [], ...options }: any) {
  const { operation, resolvedData, addValidationError } = options;

  for (const field of hasMany) await hasManyCheck({ field, ...options });

  for (const field of hasOne) {
    const hasExistingField = await hasOneExistingValue({ field, ...options });
    const noField = !resolvedData[field];
    const fieldRemoved = resolvedData[field]?.disconnect;

    if (
      (operation === 'create' && noField) ||
      (operation === 'update' && (fieldRemoved || (noField && !hasExistingField)))
    )
      addValidationError(`Missing required relationship: ${field}`);
  }
}
```

Now you can do one of the following, in a list hook:

```ts
\\ either

   validateInput: async (options) => 
      validateRelationships({ hasOne: ['config', 'eventReportingCampaign'], ...options });

\\ or

   validateInput: async (options) =>
      validateRelationships({ hasMany: ['builds'], ...options }),
```

Note that this doesn't handle a check if you made the relationship editable on the other side, like @molomby  mentioned, but you ought to be able to extend it with the same approach.

**Comment by jschuur** [comment]
Just in case anyone finds this helpful, I also wrote a little generic helper for setting a default for one to many relationships:

```js
export async function defaultRelationships({ hasOne = [], resolvedData, operation, context }) {
  if (operation !== 'create') return resolvedData;

  for (const defaultValue of hasOne) {
    const { field, targetField = 'name', targetValue } = defaultValue;

    if (resolvedData[field] || !field || !targetValue) continue;

    const targetList =
      defaultValue.targetList || field.charAt(0).toUpperCase() + field.slice(1);
    const result = await context.db[targetList].findMany({ 
      where: { [targetField]: { equals: targetValue } }
    });
    const id = result?.[0]?.id;

    if (id) resolvedData[field] = { connect: { id } };
  }

  return resolvedData;
}
```

Then in a `resolveInput` hook, I run something like this, for a `creative` field on the current list to default to   the entry in the Creative list with a name of 'Quiz System':

```js
      await defaultRelationships({
        hasOne: [{ field: 'creative', targetValue: 'Quiz System' }],
        ...options,
      });
```

I haven't needed/implement the `hasMany` equivalent, where you could set at least one default for a many to many relationship, but that would be easy to update for.

**Comment by dk-kasumov** [comment]
Hello guys! 
It is my improved solution for validating relationship fields. 
it allows to validate fields while editing & creating for many to many relations, many to one and one to one. 

```ts
async function hasManyCheck({operation, field, listKey, context, resolvedData, item, addValidationError,}) {
    console.log(operation, resolvedData[field]);
    if (operation === 'create' && !resolvedData[field]) {
        addValidationError(`Missing required relationship: ${field}`);
        return;
    }

    const currentModel = await context.prisma[listKey].findFirst({
        where: {
            id: { equals: item?.id },
        },
        select: {
            _count: {
                select: {
                    [field]: true,
                },
            },
        },
    })

    const currentCount = currentModel?._count?.[field] || 0;
    const deletedCount = resolvedData?.[field]?.disconnect?.length || 0;

    if (!(currentCount - deletedCount) && !resolvedData?.[field]?.connect?.length) {
        addValidationError(`Missing required relationship: ${field}`);
    }
}

const hasOneExistingValue = async ({ field, item, listKey, context }: any) =>
    (await context.prisma[listKey].count({
        where: {
            id: { equals: item?.id },
            [field]: { isNot: null },
        },
    })) > 0;

export async function validateRelationships({ hasOne = [], hasMany = [], ...options }: any) {
    const { operation, resolvedData, addValidationError } = options;

    for (const field of hasMany) await hasManyCheck({ field, ...options });

    for (const field of hasOne) {
        const hasExistingField = await hasOneExistingValue({ field, ...options });
        const noField = !resolvedData[field];
        const fieldRemoved = resolvedData[field]?.disconnect;

        if (
            (operation === 'create' && noField) ||
            (operation === 'update' && (fieldRemoved || (noField && !hasExistingField)))
        )
            addValidationError(`Missing required relationship: ${field}`);
    }
}
```
just copy and paste it in your code)

and usage example: 

```ts
fields: {
        author: relationship({
            ref: 'Author.posts',
            many: false,
            hooks: {
                validateInput: (options) => {
                    return validateRelationships({ hasOne: ['author'], ...options });
                }
            }
        }),
        categories: relationship({
            ref: 'PostCategory.posts',
            many: true,
            hooks: {
                validateInput: (options) => {
                    return validateRelationships({ hasMany: ['categories'], ...options });
                }
            }
        }),
    },
```

**Comment by dcousens** [comment]
Please see https://github.com/keystonejs/keystone/pull/9630 for some added support for this in the UI
